control-plane/connect-inject/peering_acceptor_controller.go

```diff
@@ -0,0 +1,259 @@
+package connectinject
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"net/http"
+	"time"
+
+	"github.com/go-logr/logr"
+	consulv1alpha1 "github.com/hashicorp/consul-k8s/control-plane/api/v1alpha1"
+	"github.com/hashicorp/consul/api"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// PeeringAcceptorController reconciles a PeeringAcceptor object.
+type PeeringAcceptorController struct {
+	client.Client
+	// ConsulClient points at the agent local to the connect-inject deployment pod.
+	ConsulClient *api.Client
+	Log          logr.Logger
+	Scheme       *runtime.Scheme
+	context.Context
+}
+
+//+kubebuilder:rbac:groups=consul.hashicorp.com,resources=peeringacceptors,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=consul.hashicorp.com,resources=peeringacceptors/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=apps,resources=secrets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apps,resources=secrets/status,verbs=get
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// PeeringAcceptor resources determine whether to generate a new peering token in Consul and store it in the backend
+// specified in the spec. If the resource doesn't exist, the peering should be deleted in Consul. If the resource
+// exists, and a peering doesn't exist in Consul, it should be created. If the resource exists, and a peering does exist
+// in Consul, it should be reconciled. If the status of the resource does not match the current state of the specified
+// secret, generate a new token and store it according to the spec.
+func (r *PeeringAcceptorController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	_ = r.Log.WithValues("peering", req.NamespacedName)
+
+	r.Log.Info("received request for PeeringAcceptor:", "name", req.Name, "ns", req.Namespace)
+
+	// Get the PeeringAcceptor resource.
+	peeringAcceptor := &consulv1alpha1.PeeringAcceptor{}
+	err := r.Client.Get(ctx, req.NamespacedName, peeringAcceptor)
+
+	// If the PeeringAcceptor resource has been deleted (and we get an IsNotFound
+	// error), we need to delete it in Consul.
+	if k8serrors.IsNotFound(err) {
+		r.Log.Info("PeeringAcceptor was deleted, deleting from Consul", "name", req.Name, "ns", req.Namespace)
+		_, err := r.deletePeering(ctx, req.Name)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
```

Comment:
So my understanding from the finalizer is that it will cause an event so that I could have access to the whole PeeringAcceptor before it is truly deleted, so that I can make sure it is deleted in Consul. I think this case is handled by the existing logic though, right? Since we only need the event's req.Name, even if deletion in Consul fails, it'll get requeued with the same name, and attempt to retry the delete in Consul which is what we want? LMK if I'm missing something!